### DISCUSSION TITLE
Overload logger methods to allow single/multiple lambda parameters to be used in garbage-free mode

### DISCUSSION THREAD

**Discussion by andriantoeff** [apache/logging-log4j2#1936] [General]
Quoting from Log4J2 [Garbage-free Steady State Logging](https://logging.apache.org/log4j/2.x/manual/garbagefree.html) page:

> **Note:** not all logging is garbage free. Specifically:
> - Logging a lambda *as a parameter* (`logger.info("lambda value is {}", () -> callExpensiveMethod())`) creates a vararg array. Logging a lambda expression by itself is garbage-free: `logger.debug(() -> callExpensiveMethod())`.

From my (user) perspective, I knowingly choose lazy logging method to log a `callExpensiveMethod()`. One of the lazy-logging method that is convenient to use is using lambda as logging parameter, e.g. `logger.debug("Large data is {}", () -> callExpensiveMethod())`.

But it seems unintuitively not recommended to use in garbage free mode.

If the reason is that it creates vararg array, why not overload the method to allow static number of parameters just like the classic method:

```
void debug(Marker marker, String message, Object p0);
...
void debug(Marker marker, String message, Object p0, Object p1, Object p2, Object p3, Object p4, Object p5, Object p6,
            Object p7, Object p8, Object p9);
```

So there could be a version for lambda parameters like this:

```
void debug(Marker marker, String message, Supplier<?> ps0);
...
void debug(Marker marker, String message, Supplier<?> ps0, Supplier<?> ps1, Supplier<?> ps2, Supplier<?> ps3,
            Supplier<?> ps4, Supplier<?> ps5, Supplier<?> ps6, Supplier<?> ps7, Supplier<?> ps8, Supplier<?> ps9);
```

What makes it not an option in Log4J2?

**Comment by rgoers** [comment]
The issue is the number of combinations. While you show a method with all Suppliers once you go beyond one you will be asked to support both Suppliers and Objects for any one parameter. This makes the number of methods huge, and we already have a ton. To be honest, a better approach is to use LogBuilder and do something like:
`logger.atDebug().withMarker(MARKER).withMessage("lambda1 is {}, param2 is {}").withArg(() -> callExpenssiveMethod0()).withArg("test").log()`
Unfortunately, LogBuilder currently doesn't provide withMessage and withArg methods. I would prefer to turn your request into that as it is more flexible and doesn't blow up the Logger interface even more.

**Reply by andriantoeff** [reply]
I now understand the concerns. Thank you for explaining!
